src/monodroid/jni/monodroid-glue.cc

```diff
@@ -1913,18 +1903,19 @@ MonodroidRuntime::setup_mono_tracing (char const* const& mono_log_mask)
 	constexpr size_t MASK_DLL_LEN = sizeof(MASK_DLL) - 1;
 	constexpr char   MASK_GC[] = "gc";
 	constexpr size_t MASK_GC_LEN = sizeof(MASK_GC) - 1;
+	constexpr char   COMMA[] = ",";
```

Comment:
What's this gain us?  I find the `","` more readable…

Reply:
I prefer `","` as well, but this is again done to gain a tiny bit of performance. Using a constant character array makes the compiler choose the `.append` overload which uses the string size computed at build time, while using a `","` means it treats the literal as a `char*` pointer and we need to use `strlen` in `append` to get its length. Also, C++ guidelines now suggest using either `const char*` or `constexpr char[]` for string literals instead of the "direct" literal, the reason being that the latter is a mutable pointer pointing to read-only memory and any attempt to write to that memory causes undefined behavior. This ambiguity should be avoided whenever possible.

Reply:
@grendello [wrote](https://github.com/xamarin/xamarin-android/pull/6159#discussion_r685758397)

> while using a `","` means it treats the literal as a `char*` pointer and we need to use `strlen` in append to get its length.

Wait, what?  For `log_mask.append (",")`, why would C++ prefer `append(const char *s)` over `const char (&s)[Size])`?

So I go about and test it:

```cpp
#include <iostream>

using namespace std;

void foo (const char* s)
{
    cout << "foo(const char*): " << s << endl;
}

template<size_t Size>
void foo (const char (&s)[Size])
{
    cout << "foo(const char (&s)[" << Size << "]): " << s << endl;
}

constexpr char LITERAL[] = "constexpr";

int main ()
{
    foo ("literal");
    foo (LITERAL);
}
```

and… it *always* prefers the `foo(const char*)` overload:

```
% c++ -std=c++20 overload.cc && ./a.out
foo(const char*): literal
foo(const char*): constexpr
```

The only way to cause the `foo (const char (&s)[Size])` overload to be used is if `foo(const char*)` *does not exist*.  When `foo(const char*)` doesn't exist, then *both* `foo("literal")` and `foo(LITERAL)` invocations hit `foo (const char (&s)[Size])`.

From this, I infer that using `COMMA` everywhere won't actually help: it's almost certainly still using the `append(const char*)` overload.

I thus further infer and ask: do we need `append(const char*)` in the first place?  Yes, we do, but perhaps we could rename it so that the template version is preferred?

Reply:
So, I know what throws it off. It's the `const char*` in the overload. I modified your sample above so that the first overload has `char*` instead of `const char*` and here's the result:

```shell
$ ./t
foo(const char (&s)[8]): literal
foo(const char (&s)[10]): constexpr
```

That is.. surprising